src/github.com/travis-ci/worker/backend/bluebox.go

```diff
@@ -0,0 +1,280 @@
+				context.LoggerFromContext(ctx).WithField("block", block).WithField("err", err).Error("could not destroy block")
+			}
+		}
+
+		if ctx.Err() == gocontext.DeadlineExceeded {
+			metrics.Mark("worker.vm.provider.bluebox.boot.timeout")
+		}
+		return nil, ctx.Err()
+	}
+}
+
+func (b *BlueBoxProvider) templateIDForLanguageGroup(language, group string) string {
+	languageMapSetting := fmt.Sprintf("LANGUAGE_MAP_%s", strings.ToUpper(language))
+	if b.cfg.IsSet(languageMapSetting) {
+		language = b.cfg.Get(languageMapSetting)
+	}
+
+	templates := b.latestTemplates()
+
+	if templateID, ok := templates[fmt.Sprintf("%s-%s", language, group)]; group != "" && ok {
+		return templateID
+	}
+
+	if templateID, ok := templates[language]; ok {
+		return templateID
+	}
+
+	if t, ok := templates["default"]; ok {
+		return t
+	}
+
+	return ""
+}
+
+func (b *BlueBoxProvider) latestTemplates() map[string]string {
+	latest := map[string]goblueboxapi.Template{}
+	latestIDs := map[string]string{}
+
+	templates, err := b.client.Templates.List()
+	if err != nil {
+		fmt.Printf("error trying to get templates: %s\n", err)
+		return nil
+	}
+
+	for _, t := range templates {
+		if t.Public || !strings.HasPrefix(t.Description, "travis-") {
+			continue
+		}
+
+		language := templateRegexp.FindStringSubmatch(t.Description)[1]
+		if _, ok := latest[language]; !ok || t.Created.After(latest[language].Created) {
+			latest[language] = t
+			latestIDs[language] = t.ID
+		}
+	}
+
+	if _, ok := latestIDs["default"]; !ok {
+		for templateName, id := range latestIDs {
+			if templateName == "ruby" {
+				latestIDs["default"] = id
+			}
+		}
+	}
+
+	return latestIDs
+}
+
+type BlueBoxInstance struct {
+	client   *goblueboxapi.Client
+	block    *goblueboxapi.Block
+	password string
+}
+
+func (i *BlueBoxInstance) sshClient(ctx gocontext.Context) (*ssh.Client, error) {
+	if len(i.block.IPs) == 0 {
+		return nil, errNoBlueBoxIP
+	}
+
+	client, err := ssh.Dial("tcp6", fmt.Sprintf("[%s]:22", i.block.IPs[0].Address), &ssh.ClientConfig{
+		User: "travis",
+		Auth: []ssh.AuthMethod{
+			ssh.Password(i.password),
+		},
+	})
+
+	if err != nil {
+		metrics.Mark("worker.vm.provider.bluebox.ssh.error")
+		context.LoggerFromContext(ctx).WithField("block", i.block).WithField("vsh_id", i.block.VSHID).WithField("err", err).Error("error connecting to SSH")
+	}
+
+	return client, err
+}
+
+func (i *BlueBoxInstance) UploadScript(ctx gocontext.Context, script []byte) error {
+	client, err := i.sshClient(ctx)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	sftp, err := sftp.NewClient(client)
+	if err != nil {
+		return err
+	}
+	defer sftp.Close()
+
+	_, err = sftp.Lstat("build.sh")
+	if err == nil {
+		return ErrStaleVM
+	}
+
+	f, err := sftp.Create("build.sh")
+	if err != nil {
+		return err
+	}
+
+	_, err = f.Write(script)
+	if err != nil {
+		return err
+	}
+
+	f, err = sftp.Create("wrapper.sh")
+	if err != nil {
+		return err
+	}
+
+	_, err = fmt.Fprintf(f, wrapperSh)
```

Comment:
love them happy accidents

![](http://humorhub.net/wp-content/uploads/2013/06/ever-make-mistakes-in-life.jpg)